dev/com.ibm.ws.kernel.service/src/com/ibm/wsspi/kernel/service/utils/AtomicServiceReference.java

```diff
@@ -29,126 +31,130 @@
+        contextRef.set(context);
     }
 
     public void deactivate(ComponentContext context) {
-        ReferenceTuple<T> previous = null;
-        ReferenceTuple<T> newTuple = null;
-
-        // Try this until we manage to set/replace the value
-        do {
-            // Get the current tuple
-            previous = tuple.get();
-
-            // Create a new tuple that clears out the context (preserve the reference to the service,
-            // as that can have attributes that should be accessible by a user between deactivate
-            // and unset
-            newTuple = new ReferenceTuple<T>(null, previous.serviceRef, null);
-
-            // Try to save the new tuple: retry if someone changed the value meanwhile
-        } while (!tuple.compareAndSet(previous, newTuple));
+        contextRef.set(null);
+        synchronized (references) {
+            // clear services located from the context that is deactivating
+            references.forEach((r) -> r.locatedService.set(null));
+        }
     }
 
     /**
      * Update service reference associated with this service.
-     * 
+     *
      * @param referenceName ServiceReference for the target service. Service references are
-     *            equal if they point to the same service registration, and are ordered by
-     *            increasing service.ranking and decreasing service.id. ServiceReferences hold
-     *            no service properties: requests/queries for properties are forwarded onto
-     *            the backing service registration.
+     *                          equal if they point to the same service registration, and are ordered by
+     *                          increasing service.ranking and decreasing service.id. ServiceReferences hold
+     *                          no service properties: requests/queries for properties are forwarded onto
+     *                          the backing service registration.
      * @return true if this is replacing a previous (non-null) service reference
      */
     public boolean setReference(ServiceReference<T> reference) {
-        ReferenceTuple<T> previous = null;
-        ReferenceTuple<T> newTuple = null;
-
-        // Try this until we either manage to make our update or we know we don't have to
-        do {
-            // Get the current tuple
-            previous = tuple.get();
-
-            // If we have something to do.. 
-            if (reference != previous.serviceRef) {
-                newTuple = new ReferenceTuple<T>(previous.context, reference, null);
-            } else {
-                break; // break out. Nothing to see here.
+        synchronized (references) {
+            // TODO not sure null should be handled
+            if (reference == null) {
+                boolean isEmpty = references.isEmpty();
+                // if null just remove all references
+                references.clear();
+                return !isEmpty;
```

Comment:
The null parameter on `setReference()` would be easy to find in method entry trace when debugging internal breaks.  If this path executes in the field, something louder would help problem determination -- e.g. debug trace that dumps an `IllegalArgumentException`.

Reply:
`com.ibm.ws.jaxrs20.ejb.components.JaxRsEJBModuleInfoBuilder.unsetWebModuleHandler(ServiceReference<ModuleHandler>)` is an example where they call `setReference(null)`

```java
    protected void unsetWebModuleHandler(ServiceReference<ModuleHandler> ref) {
        webModuleHandlerRef.setReference(null);
    }
```
The existing behavior of ASR would null out out the ref and service instance of the tuple when `null` was used, effectively removing all.  I just decided to keep the existing behavior since it seemed others depended on this.  The following are also cases of this:

JaxWsExtensionFactory.java
68: endpointPublisherManagerRef.setReference(null); 
JaxWsWebContainerManagerImpl.java
106: webModuleHandlerRef.setReference(null); 
WebEndpointPublisher.java
53: jaxWsEndpointConfiguratorManagerRef.setReference(null);